packages/kbn-config-schema/src/types/intersection_type.ts

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { ObjectType, Props, ObjectTypeOptions } from './object_type';
+
+export type IntersectionTypeOptions<T extends Props = any> = ObjectTypeOptions<T>;
+
+export class IntersectionType<
+  RTS extends Array<ObjectType<any>>,
+  T extends Props
+> extends ObjectType<T> {
+  constructor(types: RTS, options?: IntersectionTypeOptions<T>) {
+    const props = types.reduce((mergedProps, type) => {
+      Object.entries(type.getPropSchemas() as Record<string, any>).forEach(([key, value]) => {
+        if (mergedProps[key] !== undefined) {
+          throw new Error(`Duplicate key found in intersection: '${key}'`);
+        }
+        mergedProps[key as keyof T] = value;
+      });
+
+      return mergedProps;
+    }, {} as T);
```

Comment:
So, I first tried to "fork" our `UnionType` to use `alternatives(..).match('all')` instead of `alternatives(..).match('any')`, but then I realized the hard way the `alternatives('all')` is (utterly?) useless because it doesn't allow unknown properties,

so basically

```ts
// (pseudocode)
alternatives([
     schema.object({ foo: schema.string() }),
     schema.object({ bar: schema.number() }),
]).match('all')
```

Doesn't validate against `{foo: 'dolly', bar: 12}` because each type reject the other one's attribute (and forcing `unknows: allow` felt like a terrible idea that can backfire so easily).

So yeah, TLDR: Just using a thin wrapper type in top of `ObjectType` to generate the "intersection" ourselves.

One thing I didn't want to handle until we need to do so was handling the "merge" of values for duplicate keys for a intersection, so for now trying to generate the intersection of types sharing at least one key will throw an error (and I feel like it's probably very fine)

Reply:
I'm happy with that approach as it surfaces this behaviour nicely, but I'd also be OK with the last provided type taking precedence for the shared key.

Reply:
> One alternative I'd also be OK with the last provided type taking precedence for the shared key

If it's fine technically, it's a pain expressing this in term of resulting type (atm it's resulting to `Type<Output> = Type<Each & Input & Type>`, adapting the type for "each subsequent type override the keys from the previous one" would likely result in type hell - if even doable?), which is why I gave up on that one